plugins/search-backend-module-pg/src/database/DatabaseDocumentStore.ts

```diff
@@ -156,9 +156,11 @@ export class DatabaseDocumentStore implements DatabaseStore {
         const valueArray = Array.isArray(value) ? value : [value];
         const valueCompare = valueArray
           .map(v => ({ [key]: v }))
-          .map(v => JSON.stringify(v));
+          .map(v => JSON.stringify(v).toLowerCase());
         query.whereRaw(
-          `(${valueCompare.map(() => 'document @> ?').join(' OR ')})`,
+          `(${valueCompare
+            .map(() => 'lower(document::text)::jsonb @> ?')
```

Comment:
I would assume that this absolutely kills performance by breaking the index and requiring a full scan of the table every search. If that's not true I'd say it needs a comment explaining why tbh. Otherwise I think it's best to look into doing this at ingestion time instead.

Reply:
@Rugvip Thanks for quick feedback, as said in the description I believe this will be pretty performance heavy too and open for suggestions. I wanted to open the PR earlier so that we could have the discussions in the open. Looking at the document is of the datatype `jsonb`, and we want to keep it that way, I'm not sure if its possible to make it case insensitive, do you have experience doing this for columns in the catalog?

Reply:
I was hoping to be able to change the collation of the `document` column, as according to the docs it should be possible to use the standard collation, but in a nondeterministic way. 

> A deterministic collation uses deterministic comparisons, which means that it considers strings to be equal only if they consist of the same byte sequence. Nondeterministic comparison may determine strings to be equal even if they consist of different bytes. Typical situations include case-insensitive comparison, accent-insensitive comparison, as well as comparison of strings in different Unicode normal forms.

But unfortunately collations are not supported on the datatype we use, `jsonb`.

Reply:
Oh, I see catalog backend works around this by lower-casing keys/values before insertion into the `entities_search` table and before querying it.

We could take the same approach, but...
1. Lower-case all fields on a document before insertion
2. Store the original, un-lowercased document in a separate column.
3. Lower-case all fields when querying
4. Modify SQL to query on the lower-cased document but return the original document

Reply:
Do we want to proceed with this other approach? asking only for follow up 🙂

Reply:
Hi @iamEAP and @camilaibs thanks for following up with different alternatives here. @iamEAP could you link to the catalog backend, I'm interested in knowing if it stores the original, lowered-case document too? But I think your suggestion is a good way forward here, let me spend some time during the week on this!

Reply:
It's kind of all over the place, but...

- I was clued into it in [this schema migration](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/migrations/20200923104503_case_insensitivity.js), which was made in [this PR](https://github.com/backstage/backstage/pull/2586) (which is kinda overwhelmingly large)
- But you can see that the [primary table hit by the entity query is `search`](https://github.com/backstage/backstage/blob/cd3aac016f4895d0a9b91d7c7c77c9a553d2be6c/plugins/catalog-backend/src/service/NextEntitiesCatalog.ts#L87) instead of `final_entities` [when the request contains a filter](https://github.com/backstage/backstage/blob/cd3aac016f4895d0a9b91d7c7c77c9a553d2be6c/plugins/catalog-backend/src/service/NextEntitiesCatalog.ts#L161).
- Whether or not the query is based on `final_entities` or `search`, [the actual `select`](https://github.com/backstage/backstage/blob/cd3aac016f4895d0a9b91d7c7c77c9a553d2be6c/plugins/catalog-backend/src/service/NextEntitiesCatalog.ts#L166) is from a (joined) `final_entities` (which contains the originally cased data)
- And [this is where the lower casing happens on insert](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/stitching/buildEntitySearch.ts#L191) (in the entity stitching process)